File: I Foundations/2 Getting Started/MergeSort.java
import java.util.Scanner;

public class MergeSort {
    static final int max = 2147483647;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] array = new int[n];
        for(int i = 0; i < n; i++) {
            array[i] = sc.nextInt();
        }
        mergeSort(array, 0, n - 1);
        for(int i = 0; i < n; i++) {
            System.out.print(array[i] + " ");
        }

    }

    static void mergeSort(int[] array, int p, int r) {
        if(p < r) {
            int q = (int)Math.floor((p + r) / 2.0);
            mergeSort(array, p, q);
            mergeSort(array, q + 1, r);
            merge(array, p, q, r);
        }
    }

    static void merge(int[] array, int p, int q, int r) {
        int n1 = q - p + 1;
        int n2 = r - q;
        int[] L = new int[n1 + 1];
        int[] R = new int[n2 + 1];
        for(int i = 1; i <= n1; i++) {
            L[i - 1] = array[p + i - 1];
        }
        for(int j = 1; j <= n2; j++) {
            R[j - 1] = array[q + j];
        }
        L[n1] = max;
        R[n2] = max;
        int i = 0;
        int j = 0;
        for(int k = p; k <= r; k++) {
            if(L[i] <= R[j]) {
                array[k] = L[i];
                i++;
            }
            else {
                array[k] = R[j];
                j++;
            }
        }
    }
}

File: I Foundations/2 Getting Started/merge_sort.cpp
#include<stdio.h>
#include<math.h>
#include<limits.h>

void mergeSort(int array[], int start, int end);
void merge(int array[], int start, int middle, int end);

int main() {
	int len;
	scanf("%d", &len);	// 读取需要排序的数组长度
	int array[len];

	// 读取需要排序的数组
	for(int i = 0; i < len; i++) {
		scanf("%d", &array[i]);
	}

	mergeSort(array, 0, len - 1);	// 排序

	// 打印排序完成的数组
	for(int i = 0; i < len; i++) {
		printf("%d ", array[i]);
	}

	return 0;
}

void mergeSort(int array[], int start, int end) {
	if(start < end) {
		int middle = floor((start + end) / 2.0);	// 取中点
		mergeSort(array, start, middle);
		mergeSort(array, middle + 1, end);
		merge(array, start, middle, end);
	}
}

void merge(int array[], int start, int middle, int end) {
	int len1 = middle - start + 1;
	int len2 = end - middle;
	int left[len1 + 1];
	int right[len2 + 1];

	// 在两个需要合并的数组最后加入 INT_MAX
	for(int i = 0; i < len1; i++) {
		left[i] = array[start + i];
	}
	for(int i = 0; i < len2; i++) {
		right[i] = array[middle + 1 + i];		
	}
	left[len1] = INT_MAX;
	right[len2] = INT_MAX;

	// 合并两个数组
	int j = 0, k = 0;
	for(int i = start; i < end + 1; i++) {
		if(left[j] > right[k]) {
			array[i] = right[k];
			k++;
		}
		else {
			array[i] = left[j];
			j++;
		}
	}
}
